src/main/java/org/prebid/server/auction/CpmRange.java

```diff
@@ -34,24 +34,27 @@ public static String fromCpm(BigDecimal cpm, PriceGranularity priceGranularity)
      * format
      */
     public static BigDecimal fromCpmAsNumber(BigDecimal cpm, PriceGranularity priceGranularity) {
-        final BigDecimal bucketMax = priceGranularity.getBucketMax();
-        if (cpm.compareTo(bucketMax) > 0) {
-            return bucketMax;
+        final BigDecimal rangeMax = priceGranularity.getRangesMax();
+        if (cpm.compareTo(rangeMax) > 0) {
+            return rangeMax;
         }
-        return findBucketFor(cpm, priceGranularity)
-                .map(ExtPriceGranularityBucket::getIncrement)
-                .map(increment -> cpm.divide(increment, 0, RoundingMode.FLOOR).multiply(increment))
-                .orElse(null);
+        final BigDecimal increment = findRangeFor(cpm, priceGranularity.getRanges()).getIncrement();
+
+        return cpm.divide(increment, 0, RoundingMode.FLOOR).multiply(increment);
     }
 
     /**
-     * Returns bucket cpm fits in.
+     * Returns range cpm fits in.
      */
-    private static Optional<ExtPriceGranularityBucket> findBucketFor(BigDecimal cpm,
-                                                                     PriceGranularity priceGranularity) {
-        return priceGranularity.getBuckets().stream()
-                .filter(bucket -> includes(cpm, bucket.getMin(), bucket.getMax()))
-                .reduce((first, second) -> second);
+    private static ExtGranularityRange findRangeFor(BigDecimal cpm, List<ExtGranularityRange> ranges) {
+        BigDecimal min = BigDecimal.ZERO;
+        for (ExtGranularityRange range : ranges) {
+            if (includes(cpm, min, range.getMax())) {
+                return range;
+            }
+            min = range.getMax();
+        }
+        throw new IllegalArgumentException("Cpm is negative or price granularity ranges are not valid.");
```

Comment:
I'm not really comfortable to have `IllegalArgumentException` thrown here even taking into account this shouldn't happen. Previous implementation just returned null as price in this case and this is consistent with how Go version behaves in this case. Let's get back to that approach too.